Automatically Blacklist Commands that Often Conflict

tylerkaraszewski opened the pull request:
@quinthar 
cc @mcnamamj @coleaeason @righdforsa @cead22 

## Overview
This change adds a `BedrockConflictMetrics` class that keeps track of success/conflict rates for commands. It's almost entirely self-contained, so tweaking it's algorithm should be straightforward without making many changes to anything else.

`BedrockServer.cpp` needed only minor changes to support this new class. For the time being, a command needs to be explicitly enabled for multi-write (via the existing mechanism) as well as allowed by `BedrockConflictMetrics` for multi-write in order to be `process`ed in a worker thread. If everything goes smoothly, we can consider removing the explicit list of multi-write-enabled commands.

## Interface
The interface to the new class is through four static methods.

`recordConflict(const string& commandName)`: Record that a conflict occurred for a given command.

`recordSuccess(const string& commandName)`: Record that a commit succeeded for a given command.

`multiWriteOK(const string& commandName)`: Check whether a command is currently OK to multi-write.

`getMultiWriteDeniedCommands()`: Get a list of commands that are currently disabled for multi-write. This is called in the `Status` command on the master node and reported with the response from that command.

## Algorithm
The chosen algorithm is fairly simple. There are two configurable constants:

`COMMAND_COUNT`: The number of recent attempts to do a `COMMIT` for this command ("this" command here means any command with the same name/methodLine, i.e., `SetNameValuePair`) from which to keep the result. I.e., when set to `100`, we'll keep track of wether the last `100` `COMMIT`s for this command were successful, or resulted in conflicts.

`FRACTION`: The fraction of recent commands (where, "recent" means the last `COMMAND_COUNT` commands) that, when the number of conflicts exceed this, the command will not be allowed for multi-write.

The current values for `COMMAND_COUNT` and `FRACTION` are `100` and `0.10`, respectively. This means that if at any point, a given command (i.e., `setNameValuePair`) hits 10 conflicts in the last 100 attempts to commit it, it will be disabled for multi-write. It will be re-enabled for multi-write when this value drops below 10.

Note that at startup, commands are treated as though the last `COMMAND_COUNT` attempts to commit them have all been successful. I.e., they will not be blacklisted on the first call if it results in a conflict because 1 of 1 attempts to commit have resulted in conflicts, exceeding `0.10`. It will always require 10 conflicts before blacklisting a command, even if we haven't yet attempted 100 commits for that command.

## Testing
Unfortunately, with multi-write being largely non-deterministic, automated testing for this is hard. Currently, the best test we have is to watch logs for `Multi-write` while running `clustertest`. This will produce output like:
```
Jun 15 18:51:28 vagrant-ubuntu-trusty-64 bedrock: xxxxx (BedrockConflictMetrics.cpp:44) recordConflict [worker3] [info] Multi-write conflict recorded for idcollision
Jun 15 18:51:28 vagrant-ubuntu-trusty-64 bedrock: xxxxx (BedrockConflictMetrics.cpp:79) multiWriteOK [worker3] [info] Multi-write DENIED for command 'idcollision' recent conflicts: 10/100.
Jun 15 18:51:28 vagrant-ubuntu-trusty-64 bedrock: xxxxx (BedrockConflictMetrics.cpp:86) multiWriteOK [worker3] [info] Multi-write changing to DENIED for command 'idcollision' recent conflicts: 10/100, total conflicts: 26/188.
Jun 15 18:51:28 vagrant-ubuntu-trusty-64 bedrock: xxxxx (BedrockConflictMetrics.cpp:44) recordConflict [worker1] [info] Multi-write conflict recorded for idcollision
Jun 15 18:51:28 vagrant-ubuntu-trusty-64 bedrock: xxxxx (BedrockConflictMetrics.cpp:79) multiWriteOK [worker1] [info] Multi-write DENIED for command 'idcollision' recent conflicts: 11/100.
Jun 15 18:51:28 vagrant-ubuntu-trusty-64 bedrock: xxxxx (BedrockConflictMetrics.cpp:79) multiWriteOK [worker4] [info] Multi-write DENIED for command 'idcollision' recent conflicts: 11/100.
Jun 15 18:51:28 vagrant-ubuntu-trusty-64 bedrock: xxxxx (BedrockConflictMetrics.cpp:79) multiWriteOK [worker7] [info] Multi-write DENIED for command 'idcollision' recent conflicts: 11/100.
Jun 15 18:51:28 vagrant-ubuntu-trusty-64 bedrock: xxxxx (BedrockConflictMetrics.cpp:58) recordSuccess [sync] [info] Multi-write success recorded for idcollision
Jun 15 18:51:28 vagrant-ubuntu-trusty-64 bedrock: xxxxx (BedrockConflictMetrics.cpp:58) recordSuccess [sync] [info] Multi-write success recorded for idcollision
Jun 15 18:51:28 vagrant-ubuntu-trusty-64 bedrock: xxxxx (BedrockConflictMetrics.cpp:58) recordSuccess [sync] [info] Multi-write success recorded for idcollision
```

This needs to be inspected to see if it looks reasonable.

quinthar commented on BedrockConflictMetrics.cpp:
```diff
@@ -0,0 +1,103 @@
+    if (it == _conflictInfoMap.end()) {
+        // If we don't find it, we assume it's OK to multi-write.
+        SINFO("Multi-write command '" << commandName << "' not tracked in BedrockConflictMetrics. Assuming OK.");
+        result = true;
```
out of curiosity, why not just `return true;` here?

tylerkaraszewski replied:
It made more sense before I refactored this function a bit. I'll change it.

quinthar commented on BedrockConflictMetrics.h:
```diff
@@ -0,0 +1,68 @@
+#include <libstuff/libstuff.h>
+#include <bitset>
+
+class BedrockConflictMetrics {
```
This feels a little weird that there is a public class for a private data structure nobody else ever touches.  I could see a different way to encapsulate this like:

```
class BedrockConflictMetrics {
public:
    bool BedrockConflictMetrics::multiWriteOK(const string& commandName);
    ...

private:
    struct Command {
        void success();
        uint64_t _totalSuccessCount = 0;
        ...
    };

    static map<string, Command> _conflictInfoMap;
};
```

Not passionate about it.

tylerkaraszewski replied:
Eh, I started with something more like that and changed it because I thought it was similar. I'm not super passionate about it either, but I'm inclined to keep it like this so I don't risk breaking something while changing it.

coleaeason commented:
Can we make the two constants configurable? Maybe as a new param for the existing set whitelist command? So we don't have to re-deploy to change these numbers? I think it'd be really helpful to tune them as we test.

quinthar commented on BedrockConflictMetrics.cpp:
```diff
@@ -0,0 +1,103 @@
+    ++_totalSuccessCount;
+    _results.reset(_resultsPtr);
+    ++_resultsPtr;
+    _resultsPtr %= COMMAND_COUNT;
```
This is super clever.

quinthar commented on BedrockConflictMetrics.cpp:
```diff
@@ -0,0 +1,103 @@
+        // Otherwise, we check to see if it's recent conflict count is too high for multi-write.
+        auto& metric = it->second;
+        int conflicts = metric.recentConflictCount();
+        uint64_t totalAttempts = metric._totalConflictCount + metric._totalSuccessCount;
```
Why bother with `totalSuccessCount` versus just initialize the bit vector to all "success"?  This would require 10 conflicts before it gets blocked, but I think that's ok.  Just seems needlessly complex, or maybe I'm missing what problem it tries to solve.

tylerkaraszewski replied:
The bit vector *is* initialized to all success, we don't need `_totalSuccessCount`, but I thought it was interesting information to log. Notably, `_totalSuccessCount` isn't just a count of successes in the last 100 tries, but in total since we initialized this command. We use it so that when we turn a command on/off for multi-write, we'll log the number of successes and failures for the life of the process. I.e., you can get a log line that looks like:

```
Multi-write changing to DENIED for command 'idcollision' recent conflicts: 10/100, total conflicts: 26/188
```

`totalSuccessCount` is just for that `188` number.

tylerkaraszewski commented:
@quinthar - updated. Addressed your comments either with feedback or code change.

@coleaeason - I made `fraction` configurable. It's less straightforward to change `COMMAND_COUNT` because we have to resize all of our sets of recent results for every command. It also probably doesn't provide as much value as changing the fraction of commands that causes blacklisting.

You'll be able to set the fraction by setting the `autoBlacklistConflictFraction` header in the `SetCommandWhitelist` command to a value between `0.0` and `1.0`.

I think we're ready to merge.
